### PR TITLE
feat(db): initialize db with tables

### PR DESCRIPTION
**Pull request by alessandromazza98**
Supersedes https://github.com/paradigmxyz/reth/pull/13125#pullrequestreview-2478579955

@mattsse

### CODE REVIEW

File: crates/cli/commands/src/common.rs
Comment (mattsse) [inline crates/cli/commands/src/common.rs:1]: I'd to leave all of this untouched

File: crates/storage/db-api/src/table.rs [hunk 1]
```diff
@@ -100,6 +100,25 @@ pub trait Table: Send + Sync + Debug + 'static {
...
+
+impl<T: Table> TableMetadata for T {
+    fn name(&self) -> &'static str {
+        T::NAME
+    }
+
+    fn is_dupsort(&self) -> bool {
+        T::DUPSORT
+    }
+}
```
Comment (mattsse) [inline crates/storage/db-api/src/table.rs:120]: why is this necessary, isn't this just a subset of `Table`?

File: crates/storage/db/src/implementation/mdbx/mod.rs [hunk 2]
```diff
@@ -445,13 +445,13 @@ impl DatabaseEnv {
     }
 
     /// Creates all the defined tables, if necessary.
-    pub fn create_tables(&self) -> Result<(), DatabaseError> {
+    pub fn create_tables<TS: TableSet>(&self) -> Result<(), DatabaseError> {
```
Comment (mattsse) [inline crates/storage/db/src/implementation/mdbx/mod.rs:448]: this should be a new function and the `create_tables` fn calls this with the builtin tables, this way we don't have to update any other code

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/storage/db-api/src/table.rs (hunk 1 above)
Comment (alessandromazza98) [reply crates/storage/db-api/src/table.rs:120]: yes you're right, this is not needed

File: crates/storage/db/src/tables/mod.rs [hunk 3]
```diff
@@ -101,10 +101,11 @@ pub trait TableViewer<R> {
...
 pub trait TableSet {
-    /// Returns all the table names in the database.
-    fn table_names(&self) -> Vec<&'static str>;
-    /// Returns `true` if the table at the given index is a `DUPSORT` table.
-    fn is_dupsort(&self, idx: usize) -> bool;
+    /// The iterator type for the tables.
+    type TableIter: Iterator<Item = Box<dyn TableMetadata>>;
+
+    /// Returns an iterator over the tables.
+    fn tables() -> Self::TableIter;
```
Comment (mattsse) [inline crates/storage/db/src/tables/mod.rs:108]: this can just return Boxed iter of Table instead, no need for the associated type

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/storage/db/src/tables/mod.rs (hunk 3 above)
Comment (alessandromazza98) [reply crates/storage/db/src/tables/mod.rs:108]: you mean directly using `Iterator<Item = Box<dyn TableMetadata>>` as return type for the `tables` function and remove the `TableIter` associated type?

File: crates/storage/db-api/src/table.rs [hunk 4]
```diff
@@ -100,6 +100,15 @@ pub trait Table: Send + Sync + Debug + 'static {
...
 }
 
+/// Trait that provides object-safe access to the table's metadata.
+pub trait TableMetadata: Send + Sync + Debug + 'static {
+    /// The table's name.
+    fn name(&self) -> &'static str;
+
+    /// Whether the table is a `DUPSORT` table.
+    fn is_dupsort(&self) -> bool;
+}
```
Comment (mattsse) [inline crates/storage/db-api/src/table.rs:110]: this is not needed, because we already have `trait Table`?

**Review by mattsse — CHANGES_REQUESTED** [review]

File: crates/storage/db-api/src/table.rs (hunk 4 above)
Comment (alessandromazza98) [reply crates/storage/db-api/src/table.rs:110]: i don't think it's possible because `Trait` has associated `const`. I get something like this:

```
error[E0038]: the trait `Table` cannot be made into an object
   --> crates/storage/db/src/tables/mod.rs:105:48
    |
105 |     fn tables() -> Box<dyn Iterator<Item = Box<dyn Table<Key = Box<dyn Encode>, Value = Box<dyn Encode>>>>>;
    |                                                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ `Table` cannot be made into an object
    |
note: for a trait to be "dyn-compatible" it needs to allow building a vtable to allow the call to be resolvable dynamically; for more information visit <https://doc.rust-lang.org/reference/items/traits.html#object-safety>
   --> /Users/alessandromazza/work/firewall/reth/crates/storage/db-api/src/table.rs:89:11
    |
89  |     const NAME: &'static str;
    |           ^^^^ the trait cannot be made into an object because it contains this associated `const`
...
92  |     const DUPSORT: bool;
    |           ^^^^^^^ the trait cannot be made into an object because it contains this associated `const`

```

Comment (alessandromazza98) [reply crates/storage/db-api/src/table.rs:110]: or am I missing something?

Comment (mattsse) [reply crates/storage/db-api/src/table.rs:110]: I see, but then we should be able to impl this helper for all T: Table by default?

File: crates/storage/db/src/implementation/mdbx/mod.rs [hunk 5]
```diff
@@ -446,12 +446,17 @@ impl DatabaseEnv {
...
 
-        for table in Tables::ALL {
-            let flags = match table.table_type() {
-                TableType::Table => DatabaseFlags::default(),
-                TableType::DupSort => DatabaseFlags::DUP_SORT,
+        for table in TS::tables() {
+            let flags = match table.is_dupsort() {
+                true => DatabaseFlags::DUP_SORT,
+                false => DatabaseFlags::default(),
             };
```
Comment (mattsse) [inline crates/storage/db/src/implementation/mdbx/mod.rs:460]: use if-else here

File: crates/storage/db-api/src/table.rs [hunk 6]
```diff
@@ -100,6 +100,15 @@ pub trait Table: Send + Sync + Debug + 'static {
     type Value: Value;
 }
 
+/// Trait that provides object-safe access to the table's metadata.
+pub trait TableMetadata: Send + Sync + Debug + 'static {
```
Comment (mattsse) [inline crates/storage/db-api/src/table.rs:104]:
```suggestion
pub trait TableInfo: Send + Sync + Debug + 'static {
```

**Review by mattsse — CHANGES_REQUESTED** [review]
some suggestions

File: crates/storage/db-api/src/table.rs (hunk 4 above)
Comment (alessandromazza98) [reply crates/storage/db-api/src/table.rs:110]: you're right

**Review by mattsse — APPROVED** [review]
lgtm

next step would be repeating the same thing for functions that call create_tables, like

https://github.com/paradigmxyz/reth/blob/main/crates/storage/db/src/mdbx.rs#L31-L39

essentially walking up the call graph
